### PR TITLE
SONARJAVA-1750 Fail method usage in catch block (new rule RSPEC-3658)

### PR DESCRIPTION
**Pull request by axel3rd**
PR for new rule : **Unit test should throw exception instead of failing in a catch block** ([SONARJAVA-1750](https://jira.sonarsource.com/browse/SONARJAVA-1750) / [RSPEC-3658](https://jira.sonarsource.com/browse/RSPEC-3658))

### CODE REVIEW

**Comment by axel3rd** [comment]
Reviewing checks in fail (hard to execute ut/it behind proxy).

**Comment by axel3rd** [comment]
PR updated with @ganncamp & @fmallet modifications. I hope I have fulfilled the new JSON description format.
This rule is not activated by default in *sonar-way* Java profile (932 issues on *commons-beanutils* it project => could be impacting depending teams practices).
***
**Note** : The *commons-beanutils* project is using this bad practice about "assertion fail" in catch block, so to be honest, I have checked the produced *issues-report-light.html* to verify if there are no false positive, but I have produced the *its/ruling/src/test/resources/commons-beanutils/squid-S3658.json* file by reverse engineering with this snippet : 
```
Document doc = Jsoup.parse(new File("/[...]/project_issue-report", "issues-report-light.html"), "UTF-8");
Element summary = doc.select("div#summary-per-file").get(0);
long issues = 0;
System.out.println("{");
for (Element table : summary.select("table.data")) {
    System.out.println("'" + table.select("div.file_title").get(0).getElementsByTag("img").get(0).attr("title") + "':[");
    for (Element line : table.select("table.sources").get(0).select("tr.row")) {
        if (line.select("td.issues").size() > 0) {
            String id = line.id();
            String nbrLine = id.substring(id.indexOf("LV") + 2);
            System.out.println(nbrLine + ",");
            issues++;
        }
    }
    System.out.println("],");
}
System.out.println("}");
System.out.println("---");
System.out.println("Issues found : " + issues);
```
Maven dependency : 
```
<dependency>
    <groupId>org.jsoup</groupId>
    <artifactId>jsoup</artifactId>
    <version>1.9.2</version>
</dependency>
```
***
Furthermore, I promote this [orchestrator PR](https://github.com/SonarSource/orchestrator/pull/1), necessary for plugin IT tests when you are working on windows, behind an authenticated proxy and with a custom Maven installation.

File: java-checks/src/main/java/org/sonar/java/checks/AssertionFailInCatchBlockCheck.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    if (isInCatch(mit)) {
+      reportIssue(mit, "Remove this failure assertion and simply add the exception type to the method signature.");
+    }
+  }
+
+  private static boolean isInCatch(Tree tree) {
+    if (tree.parent() == null) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/AssertionFailInCatchBlockCheck.java]: This whole method can probably be refactored like : 
```
Tree parent = tree.parent();
return parent != null && (parent.is(Kind.CATCH) || isInCatch(parent));
```

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S3658_java.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
+  "title": "Unit tests should throw exceptions",
+  "status": "ready",
+  "remediation": {
+    "func": "Constant\/Issue",
+    "constantCost": "2min"
+  },
+  "sqaleSubCharac": "UNIT_TESTABILITY",
```
Comment (benzonico) [inline java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S3658_java.json]: This is not required anymore and should be removed.

File: java-checks/src/main/java/org/sonar/java/checks/AssertionFailInCatchBlockCheck.java (hunk 1 above)
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/AssertionFailInCatchBlockCheck.java]: It can also probably be improved by stopping the recursion when one of the parent is a method, no need to go up to the compilation unit.

File: its/ruling/src/test/resources/java-squid/squid-S3658.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+'org.sonarsource.java:java-squid:src/test/java/org/sonar/java/ast/visitors/FileLinesVisitorTest.java':[
+181,
+],
+'org.sonarsource.java:java-squid:src/test/java/org/sonar/java/model/TreeTokenCompletenessTest.java':[
+111,
+],
+'org.sonarsource.java:java-squid:src/test/java/org/sonar/plugins/java/api/CheckRegistrarTest.java':[
+38,
+],
+}
```
Comment (benzonico) [inline its/ruling/src/test/resources/java-squid/squid-S3658.json]: missing newline

File: its/ruling/src/test/resources/sonar-server/squid-S3658.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+'org.codehaus.sonar:sonar-server:src/test/java/org/sonar/server/db/EmbeddedDatabaseTest.java':[
+55,
+],
+'org.codehaus.sonar:sonar-server:src/test/java/org/sonar/server/debt/DebtModelPluginRepositoryTest.java':[
+99,
+],
+}
```
Comment (benzonico) [inline its/ruling/src/test/resources/sonar-server/squid-S3658.json]: missing newline

**Comment by benzonico** [comment]
Hi @axel3rd ,

Thanks for the contribution (and kudos for its quality) : see my comments in the PR to improve it further. 
We'll merge it in upcoming version 4.1 once you've fixed the small mentioned issues. 
As per json files, we usually copy the output of the failing test to provide the json expected files.

File: java-checks/src/main/java/org/sonar/java/checks/AssertionFailInCatchBlockCheck.java (hunk 1 above)
Comment (axel3rd) [inline java-checks/src/main/java/org/sonar/java/checks/AssertionFailInCatchBlockCheck.java]: > This whole method can probably be refactored like : 

Many thanks, I'm always a newbie in tree API usage :)

> It can also probably be improved by stopping the recursion when one of the parent is a method, no need to go up to the compilation unit.

Initially I left all the tree to catch this corner case (but without unit test) :
```
public class MyTest {

  static {
    try {
      File.createTempFile(null, null);
    } catch (IOException e) {
      fail(e.getMessage()); // Noncompliant {{...}}
      // Should be : throw new UnsupportedOperationException(e);
    }
  }
```
But the current code doesn't catch it => question : This use case (`fail` usage in `static` block of class) is a bad practice or not ? 
(I yes, I need to investiguate why `MethodMatcher` doesn't detect this case ...)

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S3658_java.json (hunk 2 above)
Comment (axel3rd) [inline java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S3658_java.json]: @benzonico : You speak about `sqaleSubCharac` only ?

Comment (benzonico) [inline java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S3658_java.json]: Yes, should have made that clearer in my comment.

File: java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
+import org.junit.Assert;
+import org.junit.Test;
+
+public class MyTest {
+
+  static {
+    try {
+      File.createTempFile(null, null);
+    } catch (IOException e) {
+      Assert.fail(e.getMessage()); // Compliant, corner case accepted ... and MethodMatcher doesn't pass through this case
```
Comment (axel3rd) [inline java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java]: @benzonico : Not proud about that ... corner case accepted ?.

**Comment by axel3rd** [comment]
@benzonico : Fixed, perhaps just a point to status about `Assert.fail` usage in `catch` block inside a `static` block from class (line commented)

File: java-checks/src/main/java/org/sonar/java/checks/AssertionFailInCatchBlockCheck.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    if (isInCatch(mit)) {
+      reportIssue(mit, "Remove this failure assertion and simply add the exception type to the method signature.");
+    }
+  }
+
+  private static boolean isInCatch(Tree tree) {
+    // Return false when parent is 'Kind.METHOD' precludes 'Assert.fail' usage in static block of class
+    // Corner case accepted ... and MethodMatcher doesn't pass through this case (see unit test file)
+    Tree parent = tree.parent();
+    return parent != null && !parent.is(Kind.METHOD) && (parent.is(Kind.CATCH) || isInCatch(parent));
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/AssertionFailInCatchBlockCheck.java]: A small optimization to reduce number of operations could be to replace METHOD by BLOCK (this will also handle constructors).

File: java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java (hunk 5 above)
Comment (benzonico) [inline java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java]: This is an error of semantic : write java.io.IOException and the case will be raised.

File: java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+  @Test
+  public void testMethodKoWithMessage() {
+
+    try {
+      // Some code
+    } catch (Exception e) {
+      if (true) {
+        Assert.fail(e.getMessage()); // Noncompliant {{Remove this failure assertion and simply add the exception type to the method
+                                     // signature.}}
```
Comment (benzonico) [inline java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java]: Please put those expected messages on a single line.

File: java-checks/src/main/java/org/sonar/java/checks/AssertionFailInCatchBlockCheck.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    if (isInCatch(mit)) {
+      reportIssue(mit, "Remove this failure assertion and simply add the exception type to the method signature.");
+    }
+  }
+
+  private static boolean isInCatch(Tree tree) {
+    // Return false when parent is 'Kind.METHOD' precludes 'Assert.fail' usage in static block of class
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/AssertionFailInCatchBlockCheck.java]: This comment is not necessary IMO and should be removed.

File: java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java (hunk 5 above)
Comment (benzonico) [inline java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java]: This whole test file lacks at least one precise issue location : see for instance https://github.com/SonarSource/sonar-java/blob/master/java-checks/src/test/files/checks/naming/BadMethodName.java#L5 with the [[sc=;ec=]] syntax 

This should be done for one message at least to prevent regression when rule will be improved eventually.

File: java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+      }
+    }
+
+    try {
+      // Some code
+    } catch (Exception e) {
+      org.junit.Assert.fail(e.getMessage()); // Noncompliant {{Remove this failure assertion and simply add the exception type to the method
```
Comment (benzonico) [inline java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java]: Please put those expected messages on a single line.

File: java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+                                             // signature.}}
+    }
+
+    try {
+      // Some code
+    } catch (Exception e) {
+      junit.framework.Assert.fail(e.getMessage()); // Noncompliant {{Remove this failure assertion and simply add the exception type to the
```
Comment (benzonico) [inline java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java]: Please put those expected messages on a single line.

File: java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+                                                   // method signature.}}
+    }
+
+    try {
+      // Some code
+    } catch (Exception e) {
+      org.fest.assertions.Fail.fail();// Noncompliant {{Remove this failure assertion and simply add the exception type to the method
```
Comment (benzonico) [inline java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java]: Please put those expected messages on a single line.

File: java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+      // Some code
+    } catch (Exception e) {
+      org.junit.Assert.fail(); // Noncompliant {{Remove this failure assertion and simply add the exception type to the method signature.}}
+    }
+
+    try {
+      // Some code
+    } catch (Exception e) {
+      junit.framework.Assert.fail(); // Noncompliant {{Remove this failure assertion and simply add the exception type to the method
+                                     // signature.}}
```
Comment (benzonico) [inline java-checks/src/test/files/checks/AssertionFailInCatchBlockCheck.java]: Please put those expected messages on a single line.

**Comment by benzonico** [comment]
@axel3rd see my inlined comments : fixing those properly should make your PR available for merge ! Nice work.

**Comment by axel3rd** [comment]
> This is an error of semantic : write java.io.IOException and the case will be raised.
> This comment is not necessary IMO and should be removed.

Ouch, many sorry (good catch)

> [[sc=;ec=]] syntax 

Done for all

> Please put those expected messages on a single line.

My eyes are crying when a code are not CTRL+SHIFT+F, sorry :)

> replace METHOD by BLOCK

Hummm, perhps I do not understanding something, but this code :
```
    } catch (Exception e) {
      Assert.fail();
    }
```
As this hierarchy : `EXPRESSION_STATEMENT -> BLOCK -> CATCH`, so if I replace `METHOD` by `BLOCK`, all issues are not covered by the rule ... :-(

**Comment by benzonico** [comment]
@axel3rd My bad for the remark about BLOCK vs METHOD, indeed it should not change and that's ok like this. 
So this LGTM, thanks for the contribution !
